File: day02/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	score1 := 0
	score2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		games := strings.Fields(scanner.Text())
		if len(games) != 2 {
			log.Fatalln("Invalid line")
		}

		theirMove := toMove(games[0])
		ownMove := toMove(games[1])

		score1 += winner(theirMove, ownMove) + ownMove + 1
		score2 += ownMove*3 + pickMove(theirMove, ownMove) + 1
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", score1)
	fmt.Println("Part 2:", score2)
}

func pickMove(a, b int) int {
	return (a + b + 2) % 3
}

func winner(a, b int) int {
	return ((b - a + 3 + 1) % 3) * 3
}

func toMove(k string) int {
	switch k {
	case "A", "X":
		return 0
	case "B", "Y":
		return 1
	case "C", "Z":
		return 2
	default:
		log.Fatalln("Invalid move")
		return 0
	}
}

File: day02/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let (score1, score2) = {
        let mut score1 = 0;
        let mut score2 = 0;
        for line in reader.lines() {
            let line = line?;
            let (a, b) = if let [a, b] = line.split_ascii_whitespace().collect::<Vec<_>>()[..] {
                (to_move(a)?, to_move(b)?)
            } else {
                return Err("Invalid line".into());
            };
            score1 += winner_score(a, b) + b + 1;
            score2 += b * 3 + pick_move(a, b) + 1;
        }
        (score1, score2)
    };
    println!("Part 1: {}\nPart 2: {}", score1, score2);
    Ok(())
}

fn pick_move(a: i32, b: i32) -> i32 {
    (a + b + 2) % 3
}

fn winner_score(a: i32, b: i32) -> i32 {
    ((b - a + 3 + 1) % 3) * 3
}

fn to_move(a: &str) -> Result<i32, Box<dyn std::error::Error>> {
    match a {
        "A" | "X" => Ok(0),
        "B" | "Y" => Ok(1),
        "C" | "Z" => Ok(2),
        _ => Err("Invalid move".into()),
    }
}
